Serialization format

dkeeney opened the pull request:
This provides Python access to saveToFile(file, fmt) and loadFromFile(file, fmt) with fmt as a string.  fmt is one of "BINARY", "PORTABLE", "JSON", or "XML".

dkeeney commented:
This is work in progress.
I still need to do the DateEncoder.
I probably need to update the documentation.

There is a problem where it crashes (stack is messed up) in a random place in PathTest.  There were no changes anywhere in the os section but I need to find and fix that whatever is causing it.

breznak commented on src/htm/algorithms/SDRClassifier.cpp:
```diff
@@ -67,6 +66,12 @@ PDF Classifier::infer(const SDR & pattern) const {
   return probabilities;
 }
 
+void Classifier::learn(const SDR &pattern, UInt category)
```
what do these do? Maybe add doc that it's convenience for learning a single category (if I'm right about its use?)

breznak reviewed (approved):
David, thank you for your continued work on the serialization front! 
Approved with a minor comment, when the tests pass :+1:

dkeeney commented:
When you look at my changes in Serialization.hpp you will notice that I added some overloaded functions.  pybind11 seemed to be having a problem when I had two overloaded functions and one had a default.  So I added another overloaded function without the default argument.
```
saveToFile(std::string filePath, SerializableFormat fmt)
saveToFile(std::string filePath, std::string fmt)
saveToFile(std::string filePath)
```

There is probably a way to make it work without doing this but this seemed to simplify the Python bindings.

dkeeney commented:
Another change I made was to overload Classifier::learn(SDR, label) such that the label can be an integer or a const vector reference.  Almost always it is an integer so that would probably get the most use.  But that means that the compiler did not accept the syntax `{label}` as the second argument when `label` is an integer.  Not sure how much that will break.

dkeeney commented:
To get loadFromFile( ) to work everywhere, I had to create an default constructor (one without the parameters) on all serializable objects.  Since this applies to the encoders, this means that it is possible to create an encoder without configuration which is what the loadFromFile( ) needs.    Is this going to cause any usage problems?

ctrl-z-9000-times commented:
> Another change I made was to overload Classifier::learn(SDR, label) such that the label can be an integer or a const vector reference.  [...] Not sure how much that will break.

I don't think it will break too much,
but can you add a line to the CHANGELOG anyways to document this API change?

ctrl-z-9000-times commented:
> To get loadFromFile( ) to work everywhere, I had to create an default constructor (one without the parameters) on all serializable objects. Since this applies to the encoders, this means that it is possible to create an encoder without configuration which is what the loadFromFile( ) needs. Is this going to cause any usage problems?

I don't think so, especially since you documented that the default constructors should only be used in certain ways.

dkeeney commented:
Ok, I will add something in the CHANGELOG to explain the learn( ) arguments.

breznak reviewed (approved):
looks good to me :+1: 
the serialization is a never-ending process ;)

ctrl-z-9000-times commented:
Looks good to me!
